### PR TITLE
Enable setting of transaction expiration handler in active transaction management

### PR DESCRIPTION
**Pull request by brfrn169**
## Description

This PR enables the setting of the transaction expiration handler in active transaction management.

## Related issues and/or PRs

N/A

## Changes made

- Enabled the setting of the transaction expiration handler in `ActiveTransactionManagedDistributedTransactionManager` and `ActiveTransactionManagedTwoPhaseCommitTransactionManager`

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes.
- [x] Any remaining open issues linked to this PR are documented and up-to-date (Jira, GitHub, etc.).
- [x] Tests (unit, integration, etc.) have been added for the changes.
- [x] My changes generate no new warnings.
- [x] Any dependent changes in other PRs have been merged and published.

## Additional notes (optional)

N/A

## Release notes

N/A

### CODE REVIEW

File: core/src/main/java/com/scalar/db/common/ActiveTransactionManagedDistributedTransactionManager.java [hunk 1]
```diff
@@ -40,14 +53,14 @@ public ActiveTransactionManagedDistributedTransactionManager(DatabaseConfig conf
...
-              try {
-                t.rollback();
-              } catch (Exception e) {
-                logger.warn("Rollback failed. transaction ID: {}", id, e);
-              }
+              transactionExpirationHandler.get().accept(id, t);
             });
   }
 
+  public void setTransactionExpirationHandler(BiConsumer<String, DistributedTransaction> handler) {
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/common/ActiveTransactionManagedDistributedTransactionManager.java:60]: We will be able to set a transaction expiration handler after this change.

File: core/src/main/java/com/scalar/db/common/ActiveTransactionManagedTwoPhaseCommitTransactionManager.java [hunk 2]
```diff
@@ -42,14 +55,15 @@ public ActiveTransactionManagedTwoPhaseCommitTransactionManager(DatabaseConfig c
...
-              try {
-                t.rollback();
-              } catch (Exception e) {
-                logger.warn("Rollback failed. transaction ID: {}", id, e);
-              }
+              transactionExpirationHandler.get().accept(id, t);
             });
   }
 
+  public void setTransactionExpirationHandler(
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/common/ActiveTransactionManagedTwoPhaseCommitTransactionManager.java:62]: Ditto.

File: core/src/main/java/com/scalar/db/common/ActiveTransactionManagedDistributedTransactionManager.java [hunk 3]
```diff
@@ -32,6 +34,17 @@ public abstract class ActiveTransactionManagedDistributedTransactionManager
 
   private final ActiveExpiringMap<String, ActiveTransaction> activeTransactions;
 
+  private final AtomicReference<BiConsumer<String, DistributedTransaction>>
```
Comment (feeblefakie) [inline core/src/main/java/com/scalar/db/common/ActiveTransactionManagedDistributedTransactionManager.java:37]: Just a question.
I think we also have this expiration handler in ScalarDB Cluster, but I'm wondering if we need to have it in both sides: ScalarDB core and ScalarDB Cluster.
If the core implements it, the one in ScalarDB Cluster is not necessary?

**Review by feeblefakie — APPROVED** [review]
Although I left a question, LGTM! Thank you!

File: core/src/main/java/com/scalar/db/common/ActiveTransactionManagedDistributedTransactionManager.java (hunk 1 above)
Comment (komamitsu) [reply core/src/main/java/com/scalar/db/common/ActiveTransactionManagedDistributedTransactionManager.java:60]: @brfrn169 Can you tell me how the custom handler would be like?

Comment (brfrn169) [reply core/src/main/java/com/scalar/db/common/ActiveTransactionManagedDistributedTransactionManager.java:60]: I'm going to use it for the ScalarDB Auth thing. Theoretically, we can do anything for expired transactions. For example, we can output detail logs for the expired transactions.

File: core/src/main/java/com/scalar/db/common/ActiveTransactionManagedDistributedTransactionManager.java (hunk 3 above)
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/common/ActiveTransactionManagedDistributedTransactionManager.java:37]: Sorry, I can't remember this expiration handler in ScalarDB Cluster. Please give me some more information about that?

File: core/src/main/java/com/scalar/db/common/ActiveTransactionManagedDistributedTransactionManager.java [hunk 4]
```diff
@@ -32,6 +34,17 @@ public abstract class ActiveTransactionManagedDistributedTransactionManager
...
+      transactionExpirationHandler =
+          new AtomicReference<>(
+              (id, t) -> {
+                try {
+                  t.rollback();
```
Comment (komamitsu) [inline core/src/main/java/com/scalar/db/common/ActiveTransactionManagedDistributedTransactionManager.java:42]: [minor] I think this `rollback` is always necessary and any custom expiration handler must call it. But users of `setTransactionExpirationHandler()` can pass a custom expiration handler that doesn't call `rollback` by mistake. Another option would be to add a callback like this?
```java
private final AtomicReference<BiConsumer<String, DistributedTransaction>> transactionExpirationCallback = new AtomicReference<>((id, t) -> {});
private final BiConsumer<String, DistributedTransaction>
      transactionExpirationHandler =
              (id, t) -> {
                try {
                   transactionExpirationCallback.get().accept(id, t);
                }
                catch (Throwable t) {
                  // Warning
                }
                try {
                  t.rollback();
                } catch (Exception e) {
                  logger.warn("Rollback failed. transaction ID: {}", id, e);
                }
              };
```

**Review by komamitsu — APPROVED** [review]
LGTM! 👍 

Left a minor comment, but it doesn't block this PR from being merged.

File: core/src/main/java/com/scalar/db/common/ActiveTransactionManagedDistributedTransactionManager.java (hunk 4 above)
Comment (brfrn169) [reply core/src/main/java/com/scalar/db/common/ActiveTransactionManagedDistributedTransactionManager.java:42]: Good point. I agree that `rollback` is always necessary, which is why the default handler executes `rollback`. This feature is advanced and is primarily intended for internal use. Additionally, I don't plan to document it. Therefore, if a user sets a custom handler that doesn't rollback the expired transaction, I believe it falls under the user's responsibility. Thanks.

**Pull request merged**